src/main.rs

```diff
@@ -169,42 +180,137 @@ fn main() {
     }
 }
 
+fn async_blocking_run<F, I, E>(f: F) -> Result<F::Item, F::Error>
+where
+    F: Future<Item = I, Error = E> + Send + 'static,
+    F::Item: Send + 'static,
+    F::Error: Send + 'static,
+{
+    use std::sync::Mutex;
+
+    let out = Arc::new(Mutex::new(None));
+    let out_ok = Arc::clone(&out);
+    let out_err = Arc::clone(&out);
+
+    tokio::run(
+        f.map(move |i| {
+            let mut out = out_ok.lock().unwrap();
+            *out = Some(Ok(i));
+        })
+        .map_err(move |e| {
+            let mut out = out_err.lock().unwrap();
+            *out = Some(Err(e));
+        }),
+    );
+
+    let mut res = out.lock().unwrap();
+    res.take().unwrap()
+}
+
+fn format_status_codes(e: &[CheckResultElement]) -> String {
+    use std::io::Write;
+
+    let status_codes_uniq_sorted = e
+        .iter()
+        .map(|e| StatusCode::from_u16(e.status_code()).unwrap())
+        .collect::<BTreeSet<_>>()
+        .into_iter()
+        .collect::<Vec<_>>();
+
+    if status_codes_uniq_sorted.is_empty() {
+        return "no servers found".to_string();
+    }
+
+    let mut out_buf = Vec::new();
+    let mut iter = status_codes_uniq_sorted.iter();
+
+    let sc = iter.next().unwrap();
+    write!(&mut out_buf, "{}", sc).unwrap();
+    for sc in iter {
+        write!(&mut out_buf, ", {}", sc).unwrap();
+    }
+
+    String::from_utf8(out_buf).unwrap()
+}
+
+pub fn header_from_basic_auth(auth: &Auth) -> String {
+    let mut raw_pair = auth.username.clone();
+    if let Some(ref pass) = auth.password {
+        raw_pair.push(':');
+        raw_pair.push_str(pass);
+    }
+
+    format!("Basic {}", base64::encode(&raw_pair))
+}
+
 fn check_host(target: &CanaryTarget) -> CanaryCheck {
-    let mut headers = Headers::new();
-    headers.set(UserAgent::new("rcanary/0.5.0"));
+    let latency_timer = Instant::now();
 
+    let mut headers = Vec::new();
     if let Some(ref a) = target.basic_auth {
-        headers.set(Authorization(Basic {
-            username: a.username.clone(),
-            password: a.password.clone(),
-        }))
+        headers.push((AUTHORIZATION, header_from_basic_auth(a)));
```

Comment:
Wondering if this HTTP-only stuff should be moved into `HttpCheck` later on

Reply:
yeah, I'm not sure how this will eventually end up. after tokio-ization is more complete, HttpCheck might move further away from HttpTarget, so I made it configurable in the Target for now, since auth is configured per-target.